File: backend/app/schemas/race.py
"""Race schemas."""

from datetime import date

from pydantic import Field

from app.schemas.common import BaseSchema, TimestampSchema


class RaceBase(BaseSchema):
    """Base race schema."""

    name: str = Field(..., min_length=1, max_length=100, description="レース名")
    date: date = Field(..., description="開催日")
    venue: str = Field(..., min_length=1, max_length=20, description="競馬場")
    course_type: str = Field(..., min_length=1, max_length=10, description="芝/ダート")
    distance: int = Field(..., ge=800, le=4000, description="距離(m)")
    track_condition: str | None = Field(None, max_length=10, description="馬場状態")
    weather: str | None = Field(None, max_length=10, description="天気")
    grade: str = Field(..., min_length=1, max_length=10, description="グレード")
    purse: int | None = Field(None, ge=0, description="賞金(万円)")


class RaceCreate(RaceBase):
    """Schema for creating a race."""

    pass


class RaceUpdate(BaseSchema):
    """Schema for updating a race."""

    name: str | None = Field(None, min_length=1, max_length=100)
    date: date | None = None
    venue: str | None = Field(None, min_length=1, max_length=20)
    course_type: str | None = Field(None, min_length=1, max_length=10)
    distance: int | None = Field(None, ge=800, le=4000)
    track_condition: str | None = Field(None, max_length=10)
    weather: str | None = Field(None, max_length=10)
    grade: str | None = Field(None, min_length=1, max_length=10)
    purse: int | None = Field(None, ge=0)


class RaceResponse(RaceBase, TimestampSchema):
    """Race response schema."""

    id: int
    entries_count: int | None = Field(None, description="出走頭数")


class RaceListResponse(BaseSchema):
    """Race list response schema."""

    items: list[RaceResponse]
    total: int


class RaceDetailResponse(RaceResponse):
    """Race detail response with entries."""

    from app.schemas.entry import EntryResponse

    entries: list[EntryResponse] = Field(default_factory=list)

File: frontend/src/types/race.ts
// Race types

import type { CourseType, Grade, TrackCondition } from './common';

export interface Race {
  id: number;
  name: string;
  date: string;
  venue: string;
  course_type: CourseType;
  distance: number;
  track_condition: TrackCondition | null;
  weather: string | null;
  grade: Grade;
  purse: number | null;
  entries_count: number | null;
  created_at: string;
  updated_at: string;
}

export interface RaceCreate {
  name: string;
  date: string;
  venue: string;
  course_type: CourseType;
  distance: number;
  track_condition?: TrackCondition;
  weather?: string;
  grade: Grade;
  purse?: number;
}

export interface RaceUpdate {
  name?: string;
  date?: string;
  venue?: string;
  course_type?: CourseType;
  distance?: number;
  track_condition?: TrackCondition;
  weather?: string;
  grade?: Grade;
  purse?: number;
}

export interface RaceListResponse {
  items: Race[];
  total: number;
}

// Venue list
export const VENUES = [
  '東京',
  '中山',
  '阪神',
  '京都',
  '中京',
  '新潟',
  '福島',
  '小倉',
  '札幌',
  '函館',
] as const;
